### ISSUE TITLE
Integrate the new data detection method(DC-PDD) into the repository

### ISSUE THREAD

**Issue by czguo-cs** [czguo-cs/mimir_ProblemID_1#1]
Task Description: Please integrate the new data detection method(DC-PDD)  into the target repository
You need to understand its methods based on the code implementation of DC-PDD, and refer to the implementations of other methods in this library. Combined with the structure of this library, refactor the valid parts in the original code and migrate them to this library。
DC-PDD Code:
DC-PDD/src/com_det_sco.py
`import os
import zlib
import jieba
import argparse
import numpy as np
import pickle as pkl
from tqdm import tqdm
from scipy import stats
from collections import defaultdict
from transformers import AutoTokenizer
from sklearn.metrics import auc, roc_curve

def get_arg():
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument("--tar_mod", type=str, default="pythia-2.8B")
    parser.add_argument("--data", type=str, default="WikiMIA_128")
    parser.add_argument("--max_cha", type=int, default=512, desc="text length")
    parser.add_argument("--lang", type=str, default="en")
    parser.add_argument("--a", type=str, default="0.01", desc="the hyperparameter of DC-PDD")
    arg = parser.parse_args()
    return arg

def sweep(x, score):
    fpr, tpr, _ = roc_curve(x, -score)
    acc = np.max(1-(fpr+(1-tpr))/2)
    return fpr, tpr, auc(fpr, tpr), acc

def evaluate(es, fpr_threshold=0.05):
    answers = []
    metric2predictions = defaultdict(list)
    for e in es:
        answers.append(e["label"])
        for metric in e["pred"].keys():
            metric2predictions[metric].append(e["pred"][metric])

    for metric, predictions in metric2predictions.items():
        fpr, tpr,  auc, acc = sweep(np.array(answers, dtype=bool), np.array(predictions))
        low = tpr[np.where(fpr < fpr_threshold)[0][-1]]
        print("Attack %s AUC %.4f, Accuracy %.4f, TPR@5FPR of %.4f\n" %(metric, auc, acc, low))

def cal_met(pro_dis, lang, max_cha, a):
    es = []
    for i, t in enumerate(tqdm(pro_dis)):
        e = {}
        pred = {}
        if lang == "cn":
            text = "".join(jieba.lcut(t["text"])[:max_cha])
        else:
            text = " ".join(t["text"].split()[:max_cha])

        tar_tok_num = len(tokenizer.encode(text))
        tar_ppl = np.exp(-np.mean(t["tar_prob"][:tar_tok_num]))
        pred["ppl"] = tar_ppl  # larger for nonmember

        low_tok_num = len(tokenizer.encode(text.lower()))
        low_ppl = np.exp(-np.mean(t["low_prob"][:low_tok_num]))
        pred["tar_ppl/low_ppl"] = np.log(tar_ppl)/np.log(low_ppl)  # larger for nonmember

        ref_ppl = np.exp(-np.mean(t["ref_prob"][:tar_tok_num]))
        pred["tar_ppl/ref_ppl"] = np.log(tar_ppl)/np.log(ref_ppl)  # larger for nonmember

        z_lib = len(zlib.compress(bytes(text, 'utf-8')))
        pred["tar_ppl/zlib"] = np.log(tar_ppl) / z_lib  # larger for nonmember

        k = int(tar_tok_num * 0.2)
        min_k_pro = np.sort(t["tar_prob"][:tar_tok_num])[:k]
        pred[f"min_20% prob"] = -np.mean(min_k_pro).item()  # larger for nonmember

        mu = np.array(t["tar_prob_mu"])
        sigma = np.array(t["tar_prob_sigma"])
        token_log_probs = np.array(t["tar_prob"])
        mink_plus = (token_log_probs - mu) / np.sqrt(sigma)
        k = int(len(mink_plus) * 0.2)
        min_k_plus_pro = np.sort(mink_plus)[:k]
        pred['min_20%++ prob'] = -np.mean(min_k_plus_pro).item()  # larger for nonmember

        probs = np.exp(t["tar_prob"])
        input_ids = t["input_ids"]

        # tokens with first occurance in text
        indexes = []
        current_ids = []
        for i, input_id in enumerate(input_ids):
            if input_id not in current_ids:
                indexes.append(i)
                current_ids.append(input_id)

        x_pro = probs[indexes]
        x_fre = fre_dis[input_ids[indexes]]
        ce = x_pro * np.log(1 / x_fre)
        ce[ce > a] = a
        pred["DC-PDD"] = -np.mean(ce)

        e["pred"] = pred
        e["label"] = t["label"]
        es.append(e)

    return es

if __name__ == "__main__":
    args = get_arg()
    tok_dir = "hf_model/"
    tok_path = os.path.join(tok_dir, args.tar_mod)
    tokenizer = AutoTokenizer.from_pretrained(tok_path)

    pro_dis_dir = "output/pro_dis"
    fre_dis_dir = "output/fre_dis"
    pro_dis_pat = os.path.join(pro_dis_dir, args.data, f"{args.tar_mod}.pkl")
    with open(pro_dis_pat, "rb") as f:
        pro_dis = pkl.load(f)

    fre_dis_pat = os.path.join(fre_dis_dir, f"{args.tar_mod}.pkl")
    with open(fre_dis_pat, "rb") as f:
        fre_dis = pkl.load(f)

    fre_dis_npy = np.array(fre_dis)
    fre_dis_smo = (fre_dis_npy + 1) / (sum(fre_dis_npy) + len(fre_dis_npy))

    examples = cal_met(pro_dis, args.lang, args.max_cha, args.a)
    evaluate(examples)`
DC-PDD/src/com_fre_dis.py
`import os
import gzip
import json
import torch
import logging
import argparse
import jsonlines
import pickle as pkl
from collections import defaultdict
from tqdm import tqdm
from pathlib import Path
from transformers import AutoTokenizer, AutoModelForCausalLM

logging.getLogger().setLevel(logging.INFO)
logger = logging.getLogger(__name__)

def get_arg():
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument("--ref_dat", type=str, default="C4")
    parser.add_argument("--fil_num", type=int, default=15)
    parser.add_argument("--model", type=str, default="pythia-2.8B")
    parser.add_argument("--max_tok", type=int, default=1024)
    parser.add_argument("--vob_siz", type=int, default=50304, help="50304 for pythia")
    parser.add_argument("--sav_int", type=int, default=1e4)
    arg = parser.parse_args()
    return arg

def fre_dis(ref_dat, tok, fre_dis, max_tok, k):
    """
    token frequency distribution
    ref_dat: reference dataset
    tok: tokenizer
    """
    for i, e in enumerate(tqdm(ref_dat, desc=f"{k+1} sub-dataset")):
        text = e["text"]
        input_ids = tok.encode(text)[:max_tok]
        for input_ids in input_ids:
            ran_dis[input_ids] += 1

if __name__ == "__main__":
    args = get_arg()
    logging.info(f"compute token frequency distribution for {args.model} using {args.fil_num} files of {args.ext_dat}")

    out_dir = "output/fre_dis"
    out_path = os.path.join(out_dir, args.ran_dis)
    Path(out_path).mkdir(parents=True, exist_ok=True)

    mod_dir = "hf_model/"
    mod_pat = os.path.join(mod_dir, args.model)
    tokenizer = AutoTokenizer.from_pretrained(mod_pat, trust_remote_code=True)
    fre_dis = [0] * args.vob_siz

    if args.ref_dat == "C4":
        for i in range(args.fil_num):
            iter = i
            while len(str(i)) < 5:
                i = "0" + str(i)
            fil_nam = f"c4-train.{i}-of-01024.json.gz"
            ref_dat_pat = os.path.join(args.ref_dat, fil_nam)
            with open(ref_dat_pat, "r+", encoding="utf8") as f:
                sub_dataset = gzip.open(ref_dat_pat, "rb")
                examples = []
                for example in tqdm(sub_dataset):
                    example = json.loads(example)
                    examples.append(example)
                fre_dis(examples, tokenizer, fre_dis, args.max_tok, iter)
    else:
        for i in range(args.fil_num):
            iter = i
            while len(str(i)) < 4:
                i = "0" + str(i)
            fil_nam = f"part-{i}.jsonl"
            ref_dat_pat = os.path.join(ref_dat_dir, args.ref_dat, fil_nam)
            with open(ref_dat_pat, "r+", encoding="utf8") as f:
                examples = []
                for example in tqdm(jsonlines.Reader(f)):
                    examples.append(example)
                fre_dis(examples, tokenizer, fre_dis, args.max_tok, iter)

    with open(f"{out_path}/{args.model}.pkl", "wb") as f:
        pkl.dump(fre_dis, f)`
DC-PDD/src/com_pro_dis.py
`import os
import json
import torch
import openai
import logging
import argparse
import pickle as pkl
from tqdm import tqdm
from pathlib import Path
from tensor_parallel import TensorParallelPreTrainedModel
from transformers import AutoTokenizer, AutoModelForCausalLM

logging.getLogger().setLevel(logging.INFO)

def get_arg():
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument("--tar_mod", type=str, default="pythia-2.8B")
    parser.add_argument("--ref_mod", type=str, default="pythia-70M")
    parser.add_argument("--data", type=str, default="WikiMIA_128")
    parser.add_argument("--max_tok", type=int, default=1024)
    parser.add_argument("--key_nam", type=str, default="input")
    parser.add_argument("--gpu_num", type=int, default=4)
    arg = parser.parse_args()
    return arg

def load_model(tar_mod_nam, ref_mod_nam):
    devices = []
    for i in range(args.gpu_num):
        devices.append(f"cuda:{i}")

    tar_mod = AutoModelForCausalLM.from_pretrained(tar_mod_nam, return_dict=True, trust_remote_code=True)
    tar_mod = TensorParallelPreTrainedModel(tar_mod, devices)
    tar_mod.eval()
    tar_tok = AutoTokenizer.from_pretrained(tar_mod_nam, trust_remote_code=True)

    ref_mod = AutoModelForCausalLM.from_pretrained(ref_mod_nam, return_dict=True, trust_remote_code=True)
    ref_mod = TensorParallelPreTrainedModel(ref_mod, devices)
    ref_mod.eval()
    ref_tok = AutoTokenizer.from_pretrained(ref_mod_nam, trust_remote_code=True)
    return tar_mod, ref_mod, tar_tok, ref_tok

def cal_ppl(text, model, tok):
    device = model.device
    input_ids = tok.encode(text, max_length=args.max_tok, truncation=True)
    input_ids = torch.tensor(input_ids).unsqueeze(0)
    input_ids = input_ids.to(device)

    with torch.no_grad():
        output = model(input_ids, labels=input_ids)

    logit = output[1]

    # Apply softmax to the logits to get probabilities
    prob_weight = torch.nn.functional.softmax(logit, dim=-1)[0][:-1]
    prob = torch.nn.functional.log_softmax(logit, dim=-1)[0][:-1]
    prob_mu = (prob_weight * prob).sum(-1)
    prob_sigma = (prob_weight * torch.square(prob)).sum(-1) - torch.square(prob_mu)
    input_ids = input_ids[0][1:]

    probs = prob[torch.arange(len(prob)).to(device), input_ids].tolist()
    input_ids = input_ids.tolist()
    mu = prob_mu.tolist()
    sigma = prob_sigma.tolist()
    return probs, input_ids, mu, sigma

def inference(text, label, tar_mod, ref_mod, tar_tok, ref_tok):
    response = {}
    tar_prob, input_ids, mu, sigma = cal_ppl(text, tar_mod, tar_tok)
    low_prob, _, _, _ = cal_ppl(text.lower(), tar_mod, tar_tok)
    ref_prob, _, _, _ = cal_ppl(text, ref_mod, ref_tok)

    response["input_ids"] = input_ids
    response["tar_prob"] = tar_prob
    response["low_prob"] = low_prob
    response["ref_prob"] = ref_prob
    response["tar_prob_mu"] = mu
    response["tar_prob_sigma"] = sigma
    response["text"] = text
    response["label"] = label
    return response

def tok_pro_dis(dat, key_nam, tar_mod, ref_mod, tar_tok, ref_tok):
    responses = []
    for example in tqdm(dat):
        text = example[key_nam]
        label = example["label"]
        responses.append(inference(text, label, tar_mod, ref_mod, tar_tok, ref_tok))

    return  responses

if __name__ == '__main__':
    args = get_arg()
    logging.info(f"compute token probability distribution from {args.tar_mod} on {args.data}")

    out_dir = "output/pro_dis"
    out_path = os.path.join(out_dir, args.data)
    Path(out_path).mkdir(parents=True, exist_ok=True)

    dat_dir = "data/"
    dat_pat = os.path.join(dat_dir, f"{args.data}.jsonl")
    with open(dat_pat, 'r') as f:
        dataset = [json.loads(line) for line in f]

    mod_dir = "hf_model/"
    tar_mod_name = os.path.join(mod_dir, args.tar_mod)
    ref_mod_name = os.path.join(mod_dir, args.ref_mod)
    tar_model, ref_model, tar_tokenizer, ref_tokenizer = load_model(tar_mod_name, ref_mod_name)
    pro_dis = tok_pro_dis(dataset, args.key_nam, tar_model, ref_model, tar_tokenizer, ref_tokenizer)

    with open(f"{out_path}/{args.tar_mod}.pkl", "wb") as f:
        pkl.dump(pro_dis, f)`
README.md
`This is the official repository for the paper [Pretraining Data Detection for Large Language Models: A Divergence-based Calibration Method](https://arxiv.org/abs/2409.14781) by *Weichao Zhang, Ruqing Zhang, Jiafeng Guo, Maarten de Rijke, Yixing Fan, Xueqi Cheng* <br>

## Overview

we proposes DC-PDD to improve methods that directly rely on token probabilities for pretraining data detection, which tend to misclassify non-training texts containing many common words as training texts. The key idea of DC-PDD is to calibrate the token probabilities and thereby make them more informative signals for detection. The calibration process is achieved by computing the cross-entropy (i.e., the divergence) between the token probability distribution and the token frequency distribution. To facilitate the evaluation of pretraining data detection for LLMs, we introduce a new benchmark named **PatentMIA**, specifically designed for Chinese-language pretraining data detection.

![GitHub Logo](figures/DC-PDD.png)

## PatentMIA
The PatentMIA dataset serves as a benchmark designed to evaluate pretraining data detection methods, specifically in detecting Chinese-language pretraining data from models that are open-source Chinese LLMs released between January 1, 2023 and March 1, 2024 (e.g., Qwen1.5)

The dataset contains non-training and training data:
- non-training data includes text snippets crawled from Google Patent websites published after March 1, 2024.
- training data includes text snippets crawled from Google Patent websites published before January 1, 2023.

## DC-PDD (& baselines)

We first obtain the token probability distribution by querying the LLM with the text.
```
python src/com_pro_dis.py --tar_mod <model_name> --ref_dat <data_file> --data <data_file>
```

Next, we use a large-scale publicly available corpus as a reference corpus to obtain an estimation of the token frequency distribution since an LLM’s pretraining corpus is not accessible usually. 
```
python src/com_fre_dis.py --model <model_name> --ref_dat <reference_dataset>
```

We then calibrate the token probabilities by comparing the token probability distribution to the token frequency distribution to calibrate the token probability for each token, and derive a score for pretraining data detection based on the calibrated token probabilities.
```
python src/com_det_sco.py --tar_mod <model_name> --data <data_file> --max_cha <text-length> --lang <language> --a <hyperparameter of DC-PDD>
```

## Citation

If you find this work useful, please consider citing our paper:

```bibtex
@article{zhang2024pretraining,
  title={Pretraining data detection for large language models: A divergence-based calibration method},
  author={Zhang, Weichao and Zhang, Ruqing and Guo, Jiafeng and de Rijke, Maarten and Fan, Yixing and Cheng, Xueqi},
  journal={arXiv preprint arXiv:2409.14781},
  year={2024}
}
````
